### PR TITLE
  feat(singlestore): add SingleStore Metadata Ingestion Source

### PR DESCRIPTION
**Pull request by AdalbertMemSQL**
This PR adds support for SingleStore as a metadata ingestion source in DataHub.

The implementation enables DataHub to connect to SingleStore via SQLAlchemy and ingest core metadata such as databases, schemas, tables, and columns.

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+import logging
+from typing import Any, List
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py]: 🟡 **Type Safety:** Consider replacing `Any` with `PipelineContext` from `datahub.ingestion.api.common`. Using `Any` defeats type checking and makes the contract unclear.

```python
from datahub.ingestion.api.common import PipelineContext
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    )
+    scheme: HiddenFromDocs[str] = "singlestoredb"
+
+    include_stored_procedures: bool = Field(
+        default=True,
+        description="Include ingest of stored procedures.",
+    )
+
+    procedure_pattern: AllowDenyPattern = Field(
+        default=AllowDenyPattern.allow_all(),
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py:52]: 🟡 **Documentation:** The description mentions "database.schema.procedure_name" format, but SingleStore is 2-tier (no schema layer). Consider updating to:

```python
description="Regex patterns in database.procedure_name format. "
"e.g., 'northwind.Add.*' to match procedures starting with Add in northwind database."
```

File: metadata-ingestion/tests/integration/singlestore/test_singlestore.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import pytest
```
Comment (treff7es) [inline metadata-ingestion/tests/integration/singlestore/test_singlestore.py:1]: 🟡 **Missing Unit Tests:** This connector has no unit tests in `tests/unit/`. Unit tests are important for:
- Fast feedback loop for developers
- Testing error handling paths
- Testing config validation and edge cases

Consider adding `tests/unit/test_singlestore.py` with tests for config validation and error handling scenarios.

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    @classmethod
+    def create(cls, config_dict, ctx):
+        config = SingleStoreConfig.model_validate(config_dict)
+        return cls(config, ctx)
+
+    def add_profile_metadata(self, inspector: Inspector) -> None:
+        if not self.config.is_profiling_enabled():
+            return
+        with inspector.engine.connect() as conn:
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py]: 🔴 Critical - Missing Error Handling: This method queries INFORMATION_SCHEMA.TABLE_STATISTICS and INFORMATION_SCHEMA.COLUMNAR_SEGMENTS without any try-except. Failures will crash the entire ingestion.

Impact:

Permission errors → cryptic crash
SingleStore version incompatibility → no guidance
Network issues → complete failure
Suggested fix: Wrap both queries in try-except blocks and report warnings on failure, allowing ingestion to continue with partial data.

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                    not in self.profile_metadata_info.dataset_name_to_storage_bytes
+                ):
+                    self.profile_metadata_info.dataset_name_to_storage_bytes[
+                        table_id
+                    ] = 0
+                self.profile_metadata_info.dataset_name_to_storage_bytes[table_id] += (
+                    row.UNCOMPRESSED_SIZE
+                )
+
+    def get_procedures_for_schema(
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py:132]: 🔴 Critical - Missing Error Handling: While the parent class wraps this with a generic catch-all, the method itself does not provide SingleStore-specific guidance when queries fail.

Suggested fix: Add try-except with actionable error messages telling users to set include_stored_procedures: false if they encounter permission issues.

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    - Table and column statistics via optional data profiling.
+    - Ingestion of stored procedures and their definitions.
+    - Mapping of SingleStore-specific types (e.g., JSON, VECTOR) to DataHub schema types.
+
+    Configuration options allow filtering of stored procedures and enabling/disabling profiling.
+    """
+
+    config: SingleStoreConfig
+
+    def __init__(self, config: SingleStoreConfig, ctx: Any):
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py]: Type Safety: Replace Any with PipelineContext:

from datahub.ingestion.api.common import PipelineContext

def __init__(self, config: SingleStoreConfig, ctx: PipelineContext):
This is the standard pattern across all DataHub sources.

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py (hunk 1 above)
Comment (AdalbertMemSQL) [reply metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py]: Updated

File: metadata-ingestion/tests/integration/singlestore/test_singlestore.py (hunk 3 above)
Comment (AdalbertMemSQL) [reply metadata-ingestion/tests/integration/singlestore/test_singlestore.py:1]: Updated

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py (hunk 4 above)
Comment (AdalbertMemSQL) [reply metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py]: Updated

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py (hunk 6 above)
Comment (AdalbertMemSQL) [reply metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py]: Updated

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py (hunk 5 above)
Comment (AdalbertMemSQL) [reply metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py:132]: Updated

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py (hunk 2 above)
Comment (AdalbertMemSQL) [reply metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py:52]: Updated

**Comment by AdalbertMemSQL** [comment]
@treff7es  It looks like the test failures are flaky and not related to the changes in this PR. 
When you have a moment, could you please do another review pass?

**Comment by AdalbertMemSQL** [comment]
Hi @treff7es ,

I’m really keen to get this merged, but it looks like the tests are consistently failing due to issues that seem unrelated to my changes.

When you have a moment, could you please help me move forward with this? I would really appreciate your assistance.

**Comment by treff7es** [comment]
> Hi @treff7es ,
> 
> I’m really keen to get this merged, but it looks like the tests are consistently failing due to issues that seem unrelated to my changes.
> 
> When you have a moment, could you please help me move forward with this? I would really appreciate your assistance.

@AdalbertMemSQL will check, don't worry about it

File: metadata-ingestion/setup.py [hunk 7]
```diff
@@ -1004,6 +1008,7 @@
         "mssql = datahub.ingestion.source.sql.mssql:SQLServerSource",
         "mysql = datahub.ingestion.source.sql.mysql:MySQLSource",
         "mariadb = datahub.ingestion.source.sql.mariadb.MariaDBSource",
+        "singlestore = datahub.ingestion.source.sql.singlestore.SingleStoreSource",
```
Comment (treff7es) [inline metadata-ingestion/setup.py]:
```suggestion
        "singlestore = datahub.ingestion.source.sql.singlestore:SingleStoreSource",
```

File: metadata-ingestion/tests/integration/singlestore/singlestore_to_file_no_db.yml [hunk 8]
```diff
@@ -0,0 +1,52 @@
+run_id: singlestore-test
+
+source:
+  type: singlestore
+  config:
+    username: root
+    password: example
+    host_port: localhost:53311
+    schema_pattern:
```
Comment (treff7es) [inline metadata-ingestion/tests/integration/singlestore/singlestore_to_file_no_db.yml]: In two-tier db even though schema_pattern exists but `database_pattern` should be used to be clear
```suggestion
    database_pattern:
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+)
+from datahub.metadata.com.linkedin.pegasus2avro.schema import (
+    BytesTypeClass,
+    RecordTypeClass,
+)
+
+logger = logging.getLogger(__name__)
+
+register_custom_type(JSON, RecordTypeClass)
+register_custom_type(VECTOR, BytesTypeClass)
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py:33]: Is this correct or should it be ArrayTypeClass?

Comment (AdalbertMemSQL) [reply metadata-ingestion/src/datahub/ingestion/source/sql/singlestore.py:33]: SingleStore returns vector data as a decoded blob, so BytesTypeClass seems to be a good option
